Added config file mode support for cmake find_package()

Knitschi opened the pull request:
These changes enable the googletest project to be used in hunter with

```
find_package(gtest CONFIG REQUIRED)
find_package(gmock CONFIG REQUIRED)
```

I only tested it on Windows.

Knitschi committed:
Added config file mode support for cmake find_package()

ruslo commented on CMakeLists.txt:
```diff
@@ -9,6 +10,8 @@ option(BUILD_GTEST "Builds the googletest subproject" OFF)
 #Note that googlemock target already builds googletest
 option(BUILD_GMOCK "Builds the googlemock subproject" ON)
 
+cmake_policy(SET CMP0048 NEW)
```
Since minimum version is `3.0` I don't think we need this line

Knitschi replied:
I will try what happens when I remove it. But I think I got an warning or error without it.

ruslo commented on googlemock/CMakeLists.txt:
```diff
@@ -37,8 +37,8 @@ endif()
 # Language "C" is required for find_package(Threads).
-project(gmock CXX C)
-cmake_minimum_required(VERSION 2.6.2)
+project(gmock VERSION 1.8.0 LANGUAGES CXX C )
```
I think `LANGUAGES CXX C` is redundant

Knitschi replied:
I just took this from the original project() call. My cmake knowledge is too limited to decide what is redundant here. Is it only the C or because it is included in CXX or are the options given somewhere in hunter?

ruslo replied:
> I just took this from the original project() call

Okay, I've missed that.

> My cmake knowledge is too limited to decide what is redundant here

As far as I know CMake will check C and C++ compiler by default so there is no need to set these languages explicitly. From what I know you need to set `LANGUAGES C` in cases when C++ compiler is not used or may not exist.

> are the options given somewhere in hunter?

No, Hunter will work fine with any settings.

ruslo commented on googlemock/CMakeLists.txt:
```diff
@@ -92,21 +92,121 @@ cxx_library(gmock_main
+#   * Debug: -DFOO_BAR_DEBUG=1
+#   * Release: -DFOO_BAR_DEBUG=0
+#   * other: -DFOO_BAR_DEBUG=0
+target_compile_definitions( gmock PUBLIC "GOOGLETEST_GMOCK_DEBUG=$<CONFIG:Debug>")
```
Why do we need this?

Knitschi replied:
The project builds without those lines. I remove them.

ruslo commented on googlemock/CMakeLists.txt:
```diff
@@ -92,21 +92,121 @@ cxx_library(gmock_main
+target_compile_definitions( gmock PUBLIC "GOOGLETEST_GMOCK_DEBUG=$<CONFIG:Debug>")
+target_compile_definitions( gmock_main PUBLIC "GOOGLETEST_GMOCK_MAIN_DEBUG=$<CONFIG:Debug>")
+
+include(GenerateExportHeader)
```
I don't see any changes in C++ code so I think this one not needed too

Knitschi replied:
I will remove them

ruslo commented on googlemock/CMakeLists.txt:
```diff
@@ -92,21 +92,121 @@ cxx_library(gmock_main
+#   * <prefix>/lib/libbar.a
+#   * <prefix>/lib/libbaz.a
+#   * header location after install: <prefix>/include/foo/Bar.hpp
+#   * headers can be included by C++ code `#include <foo/Bar.hpp>`
```
Need to update comments or remove them

Knitschi replied:
I will remove them

ruslo commented on googlemock/cmake/Config.cmake.in:
```diff
@@ -0,0 +1,4 @@
+@PACKAGE_INIT@
+
+include("${CMAKE_CURRENT_LIST_DIR}/@targets_export_name@.cmake")
+check_required_components("@PROJECT_NAME@")
```
See `no newline at end of file` warning. What editor you're using? I have a project with instructions to how to tune the environment to avoid such issues: http://0.readthedocs.org/en/latest/git.html#check-there-is-a-new-line-at-end-of-file, want to add more there.

Knitschi replied:
For cmake I mainly use Notepad++ and sometimes Visual Studio. I will add a newline.

ruslo replied:
> sometimes Visual Studio

Looks like it's not possible out-of-the-box: http://stackoverflow.com/questions/2254269/visual-c-setting-to-insert-a-newline-at-end-of-file

Let me know if you'll find some automatic-control solution, I'll add this info to the docs for other developers.

ruslo commented on googletest/CMakeLists.txt:
```diff
@@ -94,18 +94,126 @@ target_link_libraries(gtest_main gtest)
+#if (DEFINED CMAKE_VERSION AND NOT "${CMAKE_VERSION}" VERSION_LESS "3.0")
+#  target_include_directories(gtest      INTERFACE "${gtest_SOURCE_DIR}/include")
+#  target_include_directories(gtest_main INTERFACE "${gtest_SOURCE_DIR}/include")
+#endif()
```
I think this one must be removed

Knitschi replied:
I guess the same stuff is done by the calls to the install() function that we use. I will check if it still works when it is removed.

ruslo commented on googletest/CMakeLists.txt:
```diff
@@ -94,18 +94,126 @@ target_link_libraries(gtest_main gtest)
+  INCLUDES DESTINATION "${include_install_dir}"
+)
+
+message( "------------- targets_export_name: ${targets_export_name}")
```
I guess this is some debug printing, not needed

Knitschi replied:
Yes, I remove it.

Knitschi committed:
Removed not needed code in CMakeLists files.

Knitschi commented:
- I removed the superflous lines for the target_compile_definitions() and target_include_driectories().
- I added new lines in the config files.
- I did nothing about the LANGUAGES CXX C option that your suspected to be redundant because I am not sure what it does.

ruslo commented on googlemock/CMakeLists.txt:
```diff
@@ -92,21 +92,91 @@ cxx_library(gmock_main
+
+include(GenerateExportHeader)
+generate_export_header(gmock)
+generate_export_header(gmock_main)
```
Not needed

ruslo commented on googlemock/CMakeLists.txt:
```diff
@@ -92,21 +92,91 @@ cxx_library(gmock_main
+)
+
+# Export headers:
+install(
```
Not needed

ruslo commented on googletest/CMakeLists.txt:
```diff
@@ -91,21 +91,87 @@ cxx_library(gtest "${cxx_strict}" src/gtest-all.cc)
+############################################################################
+# Properties of targets
+
+include(GenerateExportHeader)
```
Not needed

ruslo commented on googletest/CMakeLists.txt:
```diff
@@ -91,21 +91,87 @@ cxx_library(gtest "${cxx_strict}" src/gtest-all.cc)
+)
+
+# Export headers:
+install(
```
Not needed

Knitschi committed:
- removed more unnecessary lines.

Knitschi commented:
Ok. I removed the lines that you marked.

ruslo commented on googlemock/CMakeLists.txt:
```diff
@@ -92,21 +92,91 @@ cxx_library(gmock_main
+
+#include(GenerateExportHeader)
+#generate_export_header(gmock)
+#generate_export_header(gmock_main)
```
Why comments? I think this should be just removed.

ruslo commented on googlemock/CMakeLists.txt:
```diff
@@ -92,21 +92,91 @@ cxx_library(gmock_main
+  FILES
+  "${CMAKE_CURRENT_BINARY_DIR}/gmock_export.h"
+  "${CMAKE_CURRENT_BINARY_DIR}/gmock_main_export.h"
+  DESTINATION "${include_install_dir}"
```
Should be removed

ruslo commented on googlemock/CMakeLists.txt:
```diff
@@ -92,21 +92,91 @@ cxx_library(gmock_main
+#  DIRECTORY ${gmock_SOURCE_DIR}/include/gmock
+#  DESTINATION "${include_install_dir}"
+#  FILES_MATCHING PATTERN "*.h"
+#)
```
Why headers not installed?

ruslo commented on googletest/CMakeLists.txt:
```diff
@@ -91,21 +91,87 @@ cxx_library(gtest "${cxx_strict}" src/gtest-all.cc)
+
+#include(GenerateExportHeader)
+#generate_export_header(gtest)
+#generate_export_header(gtest_main)
```
Should be removed

ruslo commented on googletest/CMakeLists.txt:
```diff
@@ -91,21 +91,87 @@ cxx_library(gtest "${cxx_strict}" src/gtest-all.cc)
+#  DIRECTORY ${gtest_SOURCE_DIR}/include/gtest
+#  DESTINATION "${include_install_dir}"
+#  FILES_MATCHING PATTERN "*.h"
+#)
```
Why headers not installed?

ruslo commented on googletest/CMakeLists.txt:
```diff
@@ -91,21 +91,87 @@ cxx_library(gtest "${cxx_strict}" src/gtest-all.cc)
+  "${CMAKE_CURRENT_BINARY_DIR}/gtest_export.h"
+  "${CMAKE_CURRENT_BINARY_DIR}/gtest_main_export.h"
+  DESTINATION "${include_install_dir}"
+)
```
Should be removed

Knitschi committed:
- removed installation of export headers
- reintroduced the installation of the normal headers

Knitschi commented:
Update

ruslo commented:
@Knitschi for compatibility reason name of the project should be `GTest` and config targets `main` and `gtest`: https://github.com/ruslo/hunter/wiki/pkg.gtest

This line can be used to "rename" gtest_main to main in config:

``` cmake
set_target_properties(gtest_main PROPERTIES EXPORT_NAME main)
```

I think it make sense for consistency reason to rename gmock to GMock.

Though probably renaming project name is too wide change, may be in make sense to rewrite just related config part.

Also just for your information I've noticed this: https://github.com/google/googletest/issues/686
